### PR TITLE
Updated API Code Samples

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Currently, the [API documentation](https://api.mattermost.com) contains a mix of PHP, Go, and Curl code examples, with the Go code referencing the very old `github.com/mattermost/mattermost-server/v5/model`, e.g.:

image

Many of these examples contain syntax errors besides not being "runnable" as-is. Remedy this with the following approach:
* Remove all PHP and Curl examples. There were only 19 Curl examples, and while there are 89 PHP examples, the community contributed [PHP API](https://github.com/gnello/php-mattermost-driver) isn't up-to-date and it's much harder to verify the quality of the examples. (We'll still link out to these kinds of libraries via the marketplace.)
* Remove all Go examples, but see below.
* Introduce the start of `ExampleClient4_*` methods in the `model_test` package that showcase valid code (organically checked as part of Go's unit testing) and render seamlessly on pkg.go.dev:

image

* Programmatically extract these examples, wrap them in a `main()` routine -- with valid imports! -- and inject `x-codeSamples` dynamically into the OpenAPIv4 spec, eventually yielding:

image

This code works! (Ok, you have to tweak the `team_id`, etc...)

I've only extracted and polished an `example_channel_test.go` file for now, and I'm not even sure there's 100% coverage there, but hopefully we can get started on recovering more examples as we go forward, perhaps even with a future linting check to ensure an example always exists for the equivalent `Client4` method.

#### Ticket Link
None.

#### Release Note
```release-note
Begin updating API examples to reflect latest Go API conventions, deprecating older code samples.
```

### CODE REVIEW

**Comment by lieut-data** [comment]
cc @cwarnermm for visibility into the proposed updates to api.mattermost.com.

**Comment by cwarnermm** [comment]
Thank you, @lieut-data!!!

**Review by BenCookie95 — APPROVED** [review]
Looks good, thank you for the detailed description @lieut-data

File: api/Makefile [hunk 1]
```diff
@@ -1,9 +1,11 @@
+ROOT := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
```
Comment (agarciamontoro) [inline api/Makefile:1]: Wow, I learned a lot of make's features with this single line. Neat!

File: api/server/main.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	"go/parser"
+	"go/printer"
+	"go/token"
+
+	"github.com/pb33f/libopenapi"
+	v3high "github.com/pb33f/libopenapi/datamodel/high/v3"
+	"golang.org/x/tools/imports"
+)
+
+// exampleText defines the template in which the corresponding ExampleClien4_* body is wrapped.
```
Comment (agarciamontoro) [inline api/server/main.go:19]: Typo
```suggestion
// exampleText defines the template in which the corresponding ExampleClient4_* body is wrapped.
```

File: api/server/main.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	applyExamples(v3Model, exampleTmpl)
+
+	// Re-render the file with the injected examples.
+	newDocument, _, _, errors := document.RenderAndReload()
+	if len(errors) > 0 {
+		for i := range errors {
+			log.Printf("error: %s\n", errors[i])
+		}
```
Comment (agarciamontoro) [inline api/server/main.go:70]: Nit levels over 9000:
```suggestion
		for _, err := range errors {
			log.Printf("error: %s\n", err)
		}
```

File: api/server/main.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	var examples []modelFunc
+	for _, pack := range packs {
+		for _, f := range pack.Files {
+			for _, d := range f.Decls {
+				if fn, isFn := d.(*ast.FuncDecl); isFn {
+					examples = append(examples, modelFunc{f, fn})
+				}
+			}
+		}
+	}
```
Comment (agarciamontoro) [inline api/server/main.go:185]: Beautiful triple loop <3

File: server/public/model/example_channel_test.go [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+		log.Fatal(err)
+	}
+	fmt.Printf("Added user %s to channel %s with roles %s\n", userId, channelId, cm.Roles)
+
+	postRootId := "post_root_id"
+	cm, _, err = client.AddChannelMemberWithRootId(context.Background(), channelId, userId, postRootId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("Added user %s to channel %s with roles %s using post %s\n", userId, channelId, cm.Roles, postRootId)
```
Comment (agarciamontoro) [inline server/public/model/example_channel_test.go:332]: Should this be its own example?

**Review by agarciamontoro — APPROVED** [review]
I *love* this, it's amazing! Thanks for this work, it's a huge improvement over what we had!

I left some comments below, half of them is just me wowing when learning new things, and the other half are just nits :)

File: server/public/model/example_channel_test.go (hunk 5 above)
Comment (lieut-data) [reply server/public/model/example_channel_test.go:333]: This second example is specially in that it has a separate `Client4` API for what amounts to the same endpoint. Usually there's just one method with optional parameters.

But in either case, you're probably right that we want to support multiple examples for a single `OperationId` (using the OpenAPI terminology). Right now, it's a 1:1, but I'd be happy to revisit that as we flesh out more examples.

Comment (agarciamontoro) [reply server/public/model/example_channel_test.go:333]: Aaah, I understand the constraints now, it's one example per endpoint. It's good as is, then, it makes sense.

File: server/public/model/example_channel_test.go [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+	client.SetToken(os.Getenv("MM_AUTHTOKEN"))
+
+	channelIds := []string{"channel_id_1", "channel_id_2"}
+
+	channels, _, err := client.GetPublicChannelsByIdsForTeam(context.Background(), "team_id", channelIds)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("Found %d channels\n", len(channels))
+}
```
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply server/public/model/example_channel_test.go]: Yeah, team_id was the important one in my opinion as well. Ok leaving the other one as is, thanks!

**Comment by lieut-data** [comment]
@cwarnermm, did you want to formally take a second pass at this now that developer review is done? (We'll continue iterating on the examples going forward.)

**Review by cwarnermm — APPROVED** [review]
Thanks, @lieut-data! I agree with removing what isn't helping developers succeed with Mattermost, and ensuring the information available is up-to-date and usable. Thank you so much for these important updates!

**Pull request merged**